iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java

```diff
@@ -0,0 +1,241 @@
+    if (!tsFileResources.isEmpty()) {
+      TsFileResource firstResource = tsFileResources.get(0);
+      TsFileID tsFileID = firstResource.getTsFileID();
+      int dataRegionId = tsFileID.regionId;
+      long partitionId = tsFileID.timePartitionId;
+      File dataRegionSysDir =
+          StorageEngine.getDataRegionSystemDir(
+              firstResource.getDatabaseName(), firstResource.getDataRegionId());
+      FileTimeIndexCacheWriter writer = getWriter(dataRegionId, partitionId, dataRegionSysDir);
+      boolean result =
+          taskQueue.offer(
+              () -> {
+                try {
+                  ByteBuffer buffer = ByteBuffer.allocate(5 * Long.BYTES * tsFileResources.size());
+                  for (TsFileResource tsFileResource : tsFileResources) {
+                    tsFileResource.serializeFileTimeIndexToByteBuffer(buffer);
+                  }
+                  buffer.flip();
+                  writer.write(buffer);
+                } catch (IOException e) {
+                  LOGGER.warn("Meet error when record FileTimeIndexCache: {}", e.getMessage());
+                }
+              });
+      if (!result) {
+        LOGGER.warn("Meet error when record FileTimeIndexCache");
+      }
+    }
+  }
+
+  public void logFileTimeIndex(TsFileResource tsFileResource) {
+    TsFileID tsFileID = tsFileResource.getTsFileID();
+    int dataRegionId = tsFileID.regionId;
+    long partitionId = tsFileID.timePartitionId;
+    File dataRegionSysDir =
+        StorageEngine.getDataRegionSystemDir(
+            tsFileResource.getDatabaseName(), tsFileResource.getDataRegionId());
+
+    FileTimeIndexCacheWriter writer = getWriter(dataRegionId, partitionId, dataRegionSysDir);
+    boolean result =
+        taskQueue.offer(
+            () -> {
+              try {
+                ByteBuffer buffer = ByteBuffer.allocate(5 * Long.BYTES);
+                tsFileResource.serializeFileTimeIndexToByteBuffer(buffer);
+                buffer.flip();
+                writer.write(buffer);
+              } catch (IOException e) {
+                LOGGER.warn("Meet error when record FileTimeIndexCache: {}", e.getMessage());
+              }
+            });
+    if (!result) {
+      LOGGER.warn("Meet error when record FileTimeIndexCache");
+    }
+  }
+
+  public void compactFileTimeIndexIfNeeded(
+      String dataBaseName,
+      int dataRegionId,
+      long partitionId,
+      TsFileResourceList sequenceFiles,
+      TsFileResourceList unsequenceFiles) {
+    FileTimeIndexCacheWriter writer =
+        getWriter(
+            dataRegionId,
+            partitionId,
+            StorageEngine.getDataRegionSystemDir(dataBaseName, String.valueOf(dataRegionId)));
+
+    int currentResourceCount =
+        (sequenceFiles == null ? 0 : sequenceFiles.size())
+            + (unsequenceFiles == null ? 0 : unsequenceFiles.size());
+    if (writer.getLogFile().length() > currentResourceCount * (5 * Long.BYTES) * 100) {
```

Comment:
Should add some comments and give the `100` a proper name.

Reply:
After the task is submitted and before it is executed, the condition is highly likely to remain true.
Is it possible that several redundant tasks are submitted during this period?